cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/HasType.java

```diff
@@ -55,27 +56,27 @@ default void setType(Type type) {
    * Set the node's type. This may start a chain of type listener notifications
    *
    * @param type new type
-   * @param root The node that initiated the type change chain. When a node receives a type setting
-   *     command where root == this, we know that we have a type listener circle and can abort. If
-   *     root == null, the type change is seen as an externally triggered event and subsequent type
-   *     listeners receive the current node as their root.
+   * @param root The nodes which we have seen in the type change chain. When a node receives a type
+   *     setting command where root.contains(this), we know that we have a type listener circle and
+   *     can abort. If root == null, the type change is seen as an externally triggered event and
+   *     subsequent type listeners receive the current node as their root.
    */
-  void setType(Type type, HasType root);
+  void setType(Type type, Collection<HasType> root);
 
   Set<Type> getPossibleSubTypes();
 
   default void setPossibleSubTypes(Set<Type> possibleSubTypes) {
-    setPossibleSubTypes(possibleSubTypes, null);
+    setPossibleSubTypes(possibleSubTypes, new ArrayList<>());
```

Comment:
aren't you checking for null in the override functions anyway, so maybe this new list is unnecessary. Otherwise, it would be good to a a non null annotation to `root`.

Reply:
Added `@NonNull`